axelrod/strategies/dbs.py

```diff
@@ -0,0 +1,398 @@
+                self.Rd.clear()
+                self.v = 0
+
+            # compute Rp for conditions that are neither in Rc or Rd
+            Rp = {}
+            all_cond = [(C, C), (C, D), (D, C), (D, D)]
+            for outcome in all_cond:
+                if ((outcome not in self.Rc.keys()) 
+                        and (outcome not in self.Rd.keys())):
+                    # then we need to compute opponent's C answer probability
+                    Rp[outcome] = self.compute_prob_rule(outcome, self.alpha)
+
+            self.Pi = {}
+            # algorithm ensure no duplicate keys -> no key overwriting
+            self.Pi.update(self.Rc)
+            self.Pi.update(self.Rd)
+            self.Pi.update(Rp)
+
+        # React to the opponent's last move
+        return MoveGen((self.history[-1], opponent.history[-1]), self.Pi,
+                depth_search_tree=self.tree_depth)
+
+
+class Node(object):
+    """
+    Nodes used to build a tree for the tree-search procedure
+    The tree has Determinist ans Stochastic nodes, as the opponent's
+    strategy is learned as a probability distribution
+    """
+
+    # abstract method
+    def get_siblings(self):
+        raise NotImplementedError('subclasses must override get_siblings()!')
+
+    # abstract method
+    def is_stochastic(self):
+        raise NotImplementedError('subclasses must override is_stochastic()!')
+
+
+class StochasticNode(Node):
+    """
+    Node that have a probability pC to get to each sibling
+    A StochasticNode can be written (C, X) or (D, X), with X = C with
+    a probability pC, else X = D
+    """
+
+    def __init__(self, own_action, pC, depth):
+        self.pC = pC
+        self.depth = depth
+        self.own_action = own_action
+
+    def get_siblings(self):
+        # siblings of a stochastic node get depth += 1 
+        opponent_c_choice = DeterministNode(self.own_action, C, self.depth+1)
+        opponent_d_choice = DeterministNode(self.own_action, D, self.depth+1)
+        return (opponent_c_choice, opponent_d_choice)
+
+    def is_stochastic(self):
+        return True
+
+
+class DeterministNode(Node):
+    """
+    Nodes (C, C), (C, D), (D, C), or (D, D) with determinist choice 
+    for siblings
+    """
+
+    def __init__(self, action1, action2, depth):
+        self.action1 = action1
+        self.action2 = action2
+        self.depth = depth
+
+    def get_siblings(self, policy):
+        """
+        build 2 siblings (C, X) and (D, X)
+        siblings of a DeterministicNode are Stochastic, and are of the
+        same depth
+        """
+        c_choice = StochasticNode(
+                C, policy[(self.action1, self.action2)], self.depth
+                )
+        d_choice = StochasticNode(
+                D, policy[(self.action1, self.action2)], self.depth
+                )
+        return (c_choice, d_choice)
+
+    def is_stochastic(self):
+        return False
+
+    def get_value(self):
```

Comment:
Suggestion: this might be better with a dict mapping a tuple of action1 and action2 to a value:
```
values = {
    (C, C): 3,
    (C, D): 0,
    (D, C): 5,
    (D, D): 1
}
return values[(action1, action2)]
```